File: TestApp/Common/ParamParser.cpp
//---------------------------------------------------------------------------
//
//	ParamParser.cpp
//
//	Implementation of parameters string parser class type.
//
//	Copyright 2017 Johnson Controls. All rights reserved.
//
//---------------------------------------------------------------------------

#include "ParamParser.hpp"
#include "Debug.hpp"

//---------------------------------------------------------------------------

	const string CParamParser::ParameterSepa = ";";
	const string CParamParser::NameValueSepa = "=";
	const string CParamParser::ValueListSepa = ",";
	const string CParamParser::OpenBracket = "[";
	const string CParamParser::CloseBracket = "]";
	const string CParamParser::WhiteSpaces = TOKENIZER_WHITE_SPACES;
	const string CParamParser::QuoteChar = string(1, TOKENIZER_QUOTE_CHAR);

	//keep this list empty forever
	const TList<string>	CParamParser::EmptyValueList;

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------

CParamParser::CParamValue::CParamValue(VOID) : IsArray(FALSE)
{
}

//---------------------------------------------------------------------------

CParamParser::CParamValue::CParamValue(CParamValue const& refValue) : IsArray(FALSE)
{
	operator =(refValue);
}

//---------------------------------------------------------------------------

CParamParser::CParamValue::~CParamValue(VOID)
{
}

//---------------------------------------------------------------------------

CParamParser::CParamValue& CParamParser::CParamValue::operator =(CParamValue const& refValue)
{
	//copy the data
	IsArray = refValue.IsArray;
	ValueText = refValue.ValueText;
	ValueList.Copy(refValue.ValueList);
	return *this;
}

//---------------------------------------------------------------------------
//---------------------------------------------------------------------------

	CParamParser::CParamParser(VOID) : 
		m_lstDictionary(new CStringComparer(TRUE))
	{
	}

//---------------------------------------------------------------------------

	CParamParser::CParamParser(string const& strParams) :
		m_lstDictionary(new CStringComparer(TRUE))
	{
		//parse the params string
		Parse(strParams);
	}

//---------------------------------------------------------------------------

	CParamParser::~CParamParser(VOID)
	{
	}

//---------------------------------------------------------------------------

	VOID CParamParser::Clear(VOID)
	{
		m_lstDictionary.Clear();
	}

//---------------------------------------------------------------------------

	BOOL CParamParser::Contains(string const& strName) const
	{
		return (FALSE != m_lstDictionary.Contains(strName))? TRUE : FALSE;
	}

//---------------------------------------------------------------------------

	TList<string> CParamParser::GetNames(VOID) const
	{
		return m_lstDictionary.GetKeys();
	}

//---------------------------------------------------------------------------

	BOOL CParamParser::IsArray(string const& strName) const
	{
		return m_lstDictionary[strName].IsArray;
	}

//---------------------------------------------------------------------------

	string const& CParamParser::GetValue(string const& strName) const
	{
		return m_lstDictionary[strName].ValueText;
	}

//---------------------------------------------------------------------------

	TList<string> const& CParamParser::GetValues(string const& strName) const
	{
		return m_lstDictionary[strName].ValueList;
	}

//---------------------------------------------------------------------------

	BOOL CParamParser::SetValue(string const& strName, 
		string const& strValue, BOOL flgOverwrite)
	{
		//create value structure
		CParamValue value;
		value.IsArray = FALSE;
		value.ValueText = strValue;

		//add value to the list
		return (NULL != m_lstDictionary.Add(
			strName, value, flgOverwrite))? TRUE : FALSE;
	}

//---------------------------------------------------------------------------

	BOOL CParamParser::SetValues(string const& strName, 
		TList<string> const& lstValues, BOOL flgOverwrite)
	{
		//create value structure
		CParamValue value;
		value.IsArray = TRUE;
		value.ValueList = lstValues;

		//add value to the list
		return (NULL != m_lstDictionary.Add(
			strName, value, flgOverwrite))? TRUE : FALSE;
	}

//---------------------------------------------------------------------------

	VOID CParamParser::Remove(string const& strName)
	{
		m_lstDictionary.Remove(strName);
	}

//---------------------------------------------------------------------------

	VOID CParamParser::Parse(string const& strParams)
	{
		//clear the current result
		Clear();
		
		//get the parameter data
		TList<string> params = CTokenizer::Split(strParams, ParameterSepa);
		size_t count = params.Count();

		//prepare the result
		string name;
		CParamValue data;

		//parse the results
		for (size_t i = 0; i < count; ++i)
		{
			//get rid of token's leading and trailing spaces
			string param = CTokenizer::Trim(params[i]);
			
			//clear param data
			name.clear();
			data.IsArray = FALSE;
			data.ValueText.clear();
			data.ValueList.Clear();
			
			//get name and value
			TList<string> pair = CTokenizer::Split(param, NameValueSepa);
			if (2 == pair.Count())	//name-value pair?
			{
				//trim and strip quotes
				name = CTokenizer::Strip(CTokenizer::Trim(pair[0]));
				
				//get the value string and try parsing as an array
				string text = CTokenizer::Trim(pair[1]);
				if (FALSE == ParseArray(text, data.ValueList))
				{
					//the text does not represent an array of values
					data.ValueText = CTokenizer::Strip(text);	//single value
				}
				else
				{
					//the param value is an array
					data.IsArray = TRUE;
				}
			}
			else if (1 == pair.Count())	//only name?
			{
				//trim and strip quotes
				name = CTokenizer::Strip(CTokenizer::Trim(pair[0]));
			}
			
			//add valid result to the table
			if (!name.empty())
			{
				//add result to the dictionary
				if (NULL == m_lstDictionary.Add(name, data, TRUE))
				{
					DEBUG_CHECK(FALSE);
					RELEASE_CHECK(FALSE);
				}
			}
		}
	}

//---------------------------------------------------------------------------

	BOOL CParamParser::ParseArray(string const& strValue, TList<string>& lstValues)
	{
		//trim the arg
		string text = CTokenizer::Trim(strValue);
		size_t size = text.length();

		//check if the value could be an array (value could contain both '[' and ']')
		if (2 <= size)
		{
			//check if the value is an array
			if (0 == text.find_first_of(OpenBracket) &&
				(size - 1 == text.find_last_of(CloseBracket)))
			{
				//the value is an array
				string list = text.substr(1, size - 2);
				lstValues = CTokenizer::Split(list, ValueListSepa);

				//trim the value strings and strip quotes
				size = lstValues.Count();
				for (size_t i = 0; i < size; ++i)
				{
					//trim and strip quotes
					lstValues[i] = 
						CTokenizer::Strip(CTokenizer::Trim(lstValues[i]));
				}

				//a list of values found
				return TRUE;
			}
		}
		//the value is not an array
		return FALSE;
	}

//---------------------------------------------------------------------------

	string CParamParser::Build(VOID) const
	{
		//preset the result
		string text;

		//iterate through the list of all params
		size_t count = m_lstDictionary.Count();
		for (size_t i = 0; i < count; ++i)
		{
			//fetch param info
			TKeyValue<string, CParamValue> const& entry = m_lstDictionary[i];

			//format the text string
			if (0 < i)	//add param separator?
			{
				text += ParameterSepa;
			}

			//append key and value
			text += FortifyText(entry.get_Key());
			text += NameValueSepa;
			text += FormatValue(entry.get_Value());
		}

		//return the result
		return text;
	}

//---------------------------------------------------------------------------

	string CParamParser::BuildArray(TList<string> const& lstValues) const
	{
		//iterate through the array
		string text;
		size_t size = lstValues.Count();
		for (size_t i = 0; i < size; ++i)
		{
			//is this the next value?
			if (0 < i)
			{
				//add separator
				text += ValueListSepa;
			}

			//add value string
			text += FortifyText(lstValues[i]);
		}

		//return the result
		return OpenBracket + text + CloseBracket;
	}

//---------------------------------------------------------------------------

	string CParamParser::FormatValue(CParamValue const& objValue) const
	{
		//is the value an array of strings?
		if (FALSE != objValue.IsArray)
		{
			//return the result
			return BuildArray(objValue.ValueList);
		}

		//the value is a string - return the result
		return FortifyText(objValue.ValueText);
	}

//---------------------------------------------------------------------------

	string CParamParser::FortifyText(string const& strValue) const
	{
		//does the string contain special chars?
		if ((string::npos != strValue.find(ParameterSepa)) ||
			(string::npos != strValue.find(NameValueSepa)) ||
			(string::npos != strValue.find(ValueListSepa)) ||
			(string::npos != strValue.find(OpenBracket)) ||
			(string::npos != strValue.find(CloseBracket)) ||
			(string::npos != strValue.find_first_of(WhiteSpaces)))
		{
			//decorate the string
			return CTokenizer::Decorate(strValue);
		}

		//return the string "as is"
		return strValue;
	}

//---------------------------------------------------------------------------

	CParamParser::CParamValue const& 
		CParamParser::operator [](string const& strName) const
	{
		return m_lstDictionary[strName];
	}

//---------------------------------------------------------------------------

File: NET_TestApp/Common/ParamParser.cs
using System;
using System.Collections.Generic;

namespace DSC.Automation.TextHelpers
{
    /// <summary>
    /// The purpose of this class is providing a tool for parsing
    /// strings containing parameters/arguments in special format.
    /// The format is: "Parameter_Name=Parameter_Value;".
    /// 
    /// The <Parameter_Name> must be one word or multiple words and/or
    /// special chars enclosed by quotes ("..."). The text enclosed
    /// in quotes may include special characters but not quote chars.
    /// 
    /// The <Parameter_Value> may be one word, multiple words enclosed 
    /// by quotes ("..."), or an <Array_Of_Values>. The text enclosed
    /// in quotes may include special characters but not quote chars.
    /// 
    /// The <Array_Of_Values> consists of a list of <List_Value> enclosed
    /// by brackets ([...]) and separated by the coma char ",".
    /// 
    /// The <List_Value> may be one word, multiple words enclosed 
    /// by quotes ("..."). The text enclosed in quotes may include 
    /// special characters but not quote chars. The <List_Value>
    /// must not be another <Array_Of_Values> - the array value nesting 
    /// is _NOT_ supported by the current implementation of the parser.
    /// 
    /// The special characters mentioned above that may be used within
    /// quoted name or value strings are: ";", "=", "[", "]", ",".
    /// Nested quotes are not supported by the parser and are not allowed
    /// anywhere in the text string to be parsed.
    /// </summary>
    public class ParamParser : MarshalByRefObject
    {
        #region Constants
        private const string _cstrParamSepa = ";";   //separates name/value pairs
        private const string _cstrNameValueSepa = "=";
        private const string _cstrValueListSepa = ",";
        private const string _cstrOpenBracket = "[";
        private const string _cstrCloseBracket = "]";
        private const string _cstrWordSepa = " \t\n";
        private const string _cstrQuoteChar = "\"";
        #endregion Constants

        #region Private Members
        private Dictionary<string, object> _lstParams =
            new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
        #endregion Private Members

        #region Properties
        public object this[string strName]
        {
            get
            {
                if (_lstParams.ContainsKey(strName))
                {
                    return _lstParams[strName];
                }
                return String.Empty;
            }
            set
            {
                if (_lstParams.ContainsKey(strName))
                {
                    //try casting value
                    string[] tblValues = value as string[];
                    string strValue = value as string;

                    //setup new value
                    if (tblValues != null)
                    {
                        _lstParams[strName] = tblValues;
                    }
                    else if (strValue != null)
                    {
                        _lstParams[strName] = strValue;
                    }
                    else
                    {
                        //set the default value
                        _lstParams[strName] = String.Empty;
                    }
                }
            }
        }
        #endregion Properties

        #region Public Methods
        public ParamParser()
        {
        }

        public override object InitializeLifetimeService()
        {
            return null;
        }

        public void Clear()
        {
            _lstParams.Clear();
        }

        public bool Contains(string strName)
        {
            return _lstParams.ContainsKey(strName);
        }

        public string[] GetNames()
        {
            List<string> lstNames = new List<string>();
            foreach (string strName in _lstParams.Keys)
            {
                lstNames.Add(strName);
            }
            return lstNames.ToArray();
        }

        public string GetValue(string strName)
        {
            if (_lstParams.ContainsKey(strName))
            {
                //return the result
                string strValue = _lstParams[strName] as string;
                if (strValue != null)
                {
                    //return the string value
                    return strValue;
                }
            }
            //return default result
            return String.Empty;
        }

        public string[] GetValues(string strName)
        {
            if (_lstParams.ContainsKey(strName))
            {
                string[] tblValues = _lstParams[strName] as string[];
                if (tblValues != null)
                {
                    //return the array of values
                    return tblValues;
                }
            }
            //return default result
            return new string[0];
        }

        //add or update if exists
        public void SetValue(string strName, string strValue)
        {
            if (_lstParams.ContainsKey(strName))
            {
                //update existing value
                _lstParams[strName] = (strValue != null) ? strValue : String.Empty;
            }
            else
            {
                //add new value
                _lstParams.Add(strName, (strValue != null) ? strValue : String.Empty);
            }
        }

        //add or update if exists
        public void SetValues(string strName, string[] tblValues)
        {
            if (_lstParams.ContainsKey(strName))
            {
                //update existing value
                _lstParams[strName] = (tblValues != null) ? tblValues : new string[0];
            }
            else
            {
                //add new value
                _lstParams.Add(strName, (tblValues != null) ? tblValues : new string[0]);
            }
        }

        //add or error if exists
        public bool AddValue(string strName, string strValue)
        {
            if (!_lstParams.ContainsKey(strName))
            {
                _lstParams.Add(strName, strValue);
                return true;
            }
            return false;
        }

        //add or error if exists
        public bool AddValues(string strName, string[] tblValues)
        {
            if (!_lstParams.ContainsKey(strName))
            {
                _lstParams.Add(strName, tblValues);
                return true;
            }
            return false;
        }

        public void Remove(string strName)
        {
            if (_lstParams.ContainsKey(strName))
            {
                _lstParams.Remove(strName);
            }
        }

        public void Parse(string strParams)
        {
            //clear the current result
            Clear();

            if (strParams != null)
            {
                //get the parameter data
                string[] tblParams = Tokenizer.Split(strParams, _cstrParamSepa);

                //parse the results
                for (int i = 0; i < tblParams.Length; ++i)
                {
                    //get rid of token's leading and trailing spaces
                    string strParam = tblParams[i].Trim();

                    //prepare the result
                    string strName = String.Empty;
                    string strValue = String.Empty; //string representation of the value
                    object objValue = null;

                    //get name and value
                    string[] tblNameValue =
                        Tokenizer.Split(strParam, _cstrNameValueSepa);
                    if (tblNameValue.Length == 2)   //pair name-value
                    {
                        //trim and strip quotes
                        strName = Tokenizer.Strip(tblNameValue[0].Trim());

                        //set the value - by default, it is the same as its string representation
                        strValue = tblNameValue[1].Trim();
                        objValue = Tokenizer.Strip(strValue);

                        //check if the value could be an array
                        string[] tblValues = ParseArray(strValue);
                        if (tblValues != null)
                        {
                            //store array as the value
                            objValue = tblValues;
                        }
                    }
                    else if (tblNameValue.Length == 1)	//only name
                    {
                        //trim and strip quotes
                        strName = Tokenizer.Strip(tblNameValue[0].Trim());
                        objValue = String.Empty;    //the value is a blank string
                    }

                    //add valid result to the table
                    if (strName != String.Empty)
                    {
                        //add result to the dictionary
                        if (!_lstParams.ContainsKey(strName))
                        {
                            //add new value
                            _lstParams.Add(strName, objValue);
                        }
                        else
                        {
                            //update existing value
                            _lstParams[strName] = objValue;
                        }
                    }
                }
            }
        }

        public string[] ParseArray(string strValue)
        {
            //trim the arg
            strValue = strValue.Trim();

            //check if the value could be an array
            if (strValue.Length >= 2)
            {
                //check if the value is an array
                if ((strValue.IndexOf(_cstrOpenBracket) == 0) &&
                    (strValue.LastIndexOf(_cstrCloseBracket) == strValue.Length - 1))
                {
                    //the value is an array
                    string strList = strValue.Substring(1, strValue.Length - 2);
                    string[] tblValues = Tokenizer.Split(strList, _cstrValueListSepa);
                    for (int v = 0; v < tblValues.Length; ++v)
                    {
                        //trim and strip quotes
                        tblValues[v] = Tokenizer.Strip(tblValues[v].Trim());
                    }

                    //return the array
                    return tblValues;
                }
            }

            //the value is not an array
            return null;
        }

        public string BuildArray(string[] tblValues)
        {
            //iterate through the array
            string strValues = String.Empty;
            for (int i = 0; i < tblValues.Length; ++i)
            {
                if (i == 0)
                {
                    strValues = FortifyText(tblValues[i]);
                }
                else
                {
                    strValues = strValues + _cstrValueListSepa +
                        FortifyText(tblValues[i]);
                }
            }

            //return the result
            return _cstrOpenBracket + strValues + _cstrCloseBracket;
        }

        public override string ToString()
        {
            //preset the result
            string strResult = String.Empty;

            //iterate through the list of all params
            bool flgFirst = true;
            foreach (KeyValuePair<string, object> clsParam in _lstParams)
            {
                if (flgFirst)
                {
                    //first param
                    flgFirst = false;
                    strResult = strResult +
                        FormatText(clsParam.Key) + _cstrNameValueSepa + FormatText(clsParam.Value);
                }
                else
                {
                    //next param (add param separator)
                    strResult = strResult + _cstrParamSepa +
                        FormatText(clsParam.Key) + _cstrNameValueSepa + FormatText(clsParam.Value);
                }
            }
            return strResult;
        }
        #endregion Public Methods

        #region Private Methods
        private string FormatText(object objValue)
        {
            //is the value an array of strings?
            if (objValue is string[])
            {
                //return the result
                return BuildArray(objValue as string[]);
            }
            else if (objValue is string)    //is the value a string?
            {
                //return the result
                return FortifyText(objValue as string);
            }

            //no valid value given
            return String.Empty;
        }

        private string FortifyText(string strValue)
        {
            //does the string contain special chars?
            if ((strValue.IndexOf(_cstrParamSepa) >= 0) ||
                (strValue.IndexOf(_cstrNameValueSepa) >= 0) ||
                (strValue.IndexOf(_cstrValueListSepa) >= 0) ||
                (strValue.IndexOf(_cstrOpenBracket) >= 0) ||
                (strValue.IndexOf(_cstrCloseBracket) >= 0) ||
                (strValue.IndexOfAny(_cstrWordSepa.ToCharArray()) >= 0))
            {
                //is the string longer than 2 chars?
                if (strValue.Length > 2)
                {
                    //is the string in quotes already?
                    if ((strValue.IndexOf(_cstrQuoteChar) == 0) &&
                        (strValue.LastIndexOf(_cstrQuoteChar) == strValue.Length - 1))
                    {
                        //the string is in quotes so there is no need to do anything
                    }
                    else
                    {
                        //add quotes to the string
                        strValue = _cstrQuoteChar + strValue + _cstrQuoteChar;
                    }
                }
                else
                {
                    //add quotes to the string
                    strValue = _cstrQuoteChar + strValue + _cstrQuoteChar;
                }
            }

            //return the string "as is"
            return strValue;
        }
        #endregion Private Methods
    }
}
